### PR TITLE
added quick and dirty LYWSD03MMC support

### PR DESCRIPTION
**Pull request by ulrich-berl**
thanks to Magalex2x14

[LYWSD03MMC-info](https://github.com/Magalex2x14/LYWSD03MMC-info)
[LYWSD03MMC support](https://github.com/custom-components/sensor.mitemp_bt/issues/7)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,7 +40,8 @@
   "dependencies": {
     "@abandonware/noble": "1.9.2-5",
     "fakegato-history": "^0.5.6",
-    "mqtt": "^3.0.0"
+    "mqtt": "^3.0.0",
+    "aes-ccm": "^1.0.3"
```
Comment (hannseman) [inline package.json]: IMO I think we should take this opportunity to move from supporting node >=8 and instead support node >=10. Node 8 LTS was dropped earlier this year. This means that we can use the native crypto module in node: https://nodejs.org/api/crypto.html

File: lib/parser.js [hunk 2]
```diff
@@ -144,6 +152,39 @@ class Parser {
     return this.buffer.readUInt8(this.eventOffset + 2);
   }
 
+  decrypt_payload() {
```
[the remark this answers is not in the document]
Comment (ulrich-berl) [reply lib/parser.js]: sure

File: package.json (hunk 1 above)
Comment (ulrich-berl) [reply package.json]: will check the native crypto module

**Comment by ulrich-berl** [comment]
> * Tests. If you could provide an advertisement data string along with a matching `bindKey` I can fix this.

bindKey: <B2D46F0CD168C18B247C0C79E9AD5B8D>
serviceData: <58585b05db184bf838c1a472c3fa42cd050000ce7b8a28>

Humidity: 43.9%

**Comment by ulrich-berl** [comment]
> * Add some documentation about how to get hold of the `bindKey`. I tried to get an understanding about the process but maybe you have some more details?

Best check the comments in [LYWSD03MMC support](https://github.com/custom-components/sensor.mitemp_bt/issues/7):

**via Charles proxy** _"Bottom line: to decrypt adv messages, we only need the key (bind_key or beaconkey, individual for every sensor) that I captured using the Charles proxy application on my iphone (from the MiHome application traffic during "pairing" with the sensor, no jail needed). This key can also be obtained from the Internet traffic of Xiaomi BT-gateway. I don’t know any other ways. In principle, the procedure is not complicated, but maybe someone know an easier option."_

**via MiHome Mod** _"My mod lives [here](https://www.kapiba.ru/2017/11/mi-home.html), new version with bind_key logging already published :) Just install the mod, create folder /your_interlal_storage/vevs/logs/ and pair a device - file pairings.txt with corresponding data will be created."_

**via ios app stream-network-debug-tool** _"Make sure you enable HTTPS Sniffing and install the certificate otherwise you won't get the values. I then exported everything (a har file) and opened that with Charles Proxy and followed the rest of your suggestions."_

**Comment by ulrich-berl** [comment]
maybe there should be an additional hasEvent check in scanner.js before evaluate the eventtype...

```
if (!result.frameControl.hasEvent) {
  return;
}
```

**Comment by hannseman** [comment]
Awesome! I fixed a few things, hope you don't mind:

* I added a check if `bindKey` is missing when encountering encrypted payloads.
* Added documentation for the `bindKey` option.
* Removed support for Node 8 in CI (I'll need to remove the tests on GitHub Actions manually)

**Comment by hannseman** [comment]
> maybe there should be an additional hasEvent check in scanner.js before evaluate the eventtype...

Yes I guess that makes sense. Do you know in what cases `hasEvent` is set to false?

```javascript
if (!result.frameControl.hasEvent) {
  this.log.debug("No event");
  return;
}
```

**Comment by ulrich-berl** [comment]
> Awesome! I fixed a few things, hope you don't mind:
> 
> * I added a check if `bindKey` is missing when encountering encrypted payloads.
> * Added documentation for the `bindKey` option.
> * Removed support for Node 8 in CI (I'll need to remove the tests on GitHub Actions manually)

👍

**Comment by ulrich-berl** [comment]
> > maybe there should be an additional hasEvent check in scanner.js before evaluate the eventtype...
> 
> Yes I guess that makes sense. Do you know in what cases `hasEvent` is set to false?
> 
> ```js
> if (!result.frameControl.hasEvent) {
>   this.log.debug("No event");
>   return;
> }
> ```

if you check [LYWSD03MMC-info](https://github.com/Magalex2x14/LYWSD03MMC-info) there are frame (unbinded/08 payload) having no data (see data flag), so event flag in frameControl will be false.

**Comment by hannseman** [comment]
Just merged this to master. Thanks again for your contribution!

 I'll create a new release ASAP, just need to clean up some documentation.

**Comment by ulrich-berl** [comment]
> Just merged this to master. Thanks again for your contribution!
> 
> I'll create a new release ASAP, just need to clean up some documentation.

Thank you, too !

**Pull request merged**